xcm/xcm-emulator/src/lib.rs

```diff
@@ -816,16 +833,26 @@ macro_rules! __impl_test_ext_for_parachain {
 							<$name>::send_bridged_messages(msg);
 						}
 
-						// clean messages
-						<Self as Parachain>::ParachainSystem::on_initialize(block_number);
-
 						// log events
 						Self::events().iter().for_each(|event| {
 							$crate::log::debug!(target: concat!("events::", stringify!($name)), "{:?}", event);
 						});
 
+						// Store parent head data for use later.
+						let created_header = <Self as Chain>::System::finalize();
+						let parent_head_data = $crate::LAST_HEAD.with(|b| b.borrow_mut()
+							.get_mut(<Self as NetworkComponent>::Network::name())
+							.expect("network not initialized?")
+							.insert(para_id.into(), $crate::HeadData(created_header.encode()))
+						);
+
 						// clean events
 						<Self as Chain>::System::reset_events();
+
+						// reinitialize before next call.
+						let next_block_number = block_number + 1;
+						<Self as Chain>::System::initialize(&next_block_number, &created_header.hash(), &Default::default());
+						<Self as Parachain>::ParachainSystem::on_initialize(next_block_number);
```

Comment:
It would be nice if we can just do `AllPalletsWithSystem::on_initialize` here, but it seems to not be supported currently.  
Some pallets (like the MessageQueue) only advance on initialize, and injecting them manually into the config is also ugly.

Reply:
Found a better solution without on_initialize needs.